### PR TITLE
feat: add sqs handler

### PR DESCRIPTION
**Pull request by pevisscher**

### CODE REVIEW

**Comment by pevisscher** [comment]
Current dependencies on/for this PR:
* main
  * **PR #11** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/11" target="_blank">Graphite</a> 
    * **PR #12** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/12" target="_blank">Graphite</a> 
      * **PR #14** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/14" target="_blank">Graphite</a> 
        * **PR #13** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/13" target="_blank">Graphite</a> 
          * **PR #15** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/15" target="_blank">Graphite</a> 
            * **PR #16** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/16" target="_blank">Graphite</a> 
              * **PR #17** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/17" target="_blank">Graphite</a> 
                * **PR #18** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/18" target="_blank">Graphite</a> 
                  * **PR #19** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/19" target="_blank">Graphite</a> 
                    * **PR #20** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/20" target="_blank">Graphite</a> 
                      * **PR #21** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/21" target="_blank">Graphite</a> 
                        * **PR #22** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/22" target="_blank">Graphite</a> 
                          * **PR #23** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/23" target="_blank">Graphite</a>  👈
                            * **PR #24** <a href="https://app.graphite.dev/github/pr/skyleague/event-horizon/24" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/skyleague/event-horizon/23?utm_source=stack-comment).

File: src/lib/events/sqs/handler.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            maybeFailure = errorHandlerFn.onError(event, e)
+        }
+        ioLoggerFn.after(undefined, item)
+
+        if (isJust(maybeFailure)) {
+            failures ??= []
+            failures.push(maybeFailure)
+        }
+    }
+    if (failures !== undefined) {
```
Comment (MickVanDuijn) [inline src/lib/events/sqs/handler.ts:55]:
```suggestion
    if (failures.length > 0) {
```

File: src/lib/events/sqs/handler.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        throw EventError.notImplemented()
+    }
+    const { sqs } = handler
+
+    const errorHandlerFn = sqsErrorHandler(context)
+    const parseEventFn = sqsParseEvent(sqs)
+    const ioValidateFn = ioValidate<SQSEvent>({ input: (e) => e.payload })
+    const ioLoggerFn = ioLogger({ type: 'firehose' }, context)
+
+    let failures: SQSBatchItemFailure[] | undefined = undefined
```
Comment (MickVanDuijn) [inline src/lib/events/sqs/handler.ts:30]:
```suggestion
    let failures: SQSBatchItemFailure[] = []
```

File: src/lib/events/sqs/handler.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            ioLoggerFn.before(sqsEvent, item)
+
+            await sqs.handler(sqsEvent.right, context)
+        } catch (e: unknown) {
+            maybeFailure = errorHandlerFn.onError(event, e)
+        }
+        ioLoggerFn.after(undefined, item)
+
+        if (isJust(maybeFailure)) {
+            failures ??= []
```
Comment (MickVanDuijn) [inline src/lib/events/sqs/handler.ts:51]:
```suggestion
```

File: src/lib/events/sqs/types.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+export interface SQSEvent<P = unknown> {
+    payload: P
+    raw: SQSRecord
+}
+
+export interface SQSEventHandler<C = never, S = never, P = unknown> {
+    schema: {
+        payload?: Schema<P>
+    }
+    handler: (request: P, context: LambdaContext<C, S>) => Promise<void> | void
```
Comment (MickVanDuijn) [inline src/lib/events/sqs/types.ts:16]:
```suggestion
    handler: (payload: P, context: LambdaContext<C, S>) => Promise<void> | void
```

**Review by MickVanDuijn — CHANGES_REQUESTED** [review]

File: src/lib/events/sqs/handler.ts (hunk 3 above)
Comment (pevisscher) [reply src/lib/events/sqs/handler.ts:51]: Kept verbose on purpose for debugging purposes

File: src/lib/events/sqs/handler.ts (hunk 1 above)
Comment (pevisscher) [reply src/lib/events/sqs/handler.ts:55]: Kept verbose on purpose for debugging purposes

**Review by MickVanDuijn — APPROVED** [review]

**Comment by pevisscher** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/skyleague/event-horizon/23?utm_source=stack-comment).


Job Reference: 74QS0D26BjCoc1fjew09

**Comment by pevisscher** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/skyleague/event-horizon/23?utm_source=stack-comment).


Job Reference: MNQ6fIHHP2CutIWmkx44

**Comment by pevisscher** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/skyleague/event-horizon/23?utm_source=stack-comment).


Job Reference: ZixEesAXj0mMWpqHV7EA

**Pull request merged**